onnxscript/values.py

```diff
@@ -603,7 +624,103 @@ def to_model_proto(self, **kwargs):
+                Used to set value_info for intermediate variables.
+            opset_version: The standard opset version to use for the model if it
+                cannot be inferred. Otherwise defaults to the current opset version.
+            kwargs: Additional parameters given to function :func:`onnx.helper.make_model`.
+
+        Returns:
+            An instance of :class:`onnx.ModelProto`.
+        """
+        value_infos = (
+            [
+                onnx.helper.make_value_info(name, type.to_type_proto())
+                for name, type in value_infos.items()
+            ]
+            if value_infos
+            else None
+        )
+
+        graph = self.function_ir.to_graph_proto()
+        if value_infos:
+            graph.value_info.extend(value_infos)
+        if io_types is not None:
+            for input in graph.input:
+                if not input.HasField("type"):
+                    input.type.CopyFrom(io_types.to_type_proto())
+            for output in graph.output:
+                if not output.HasField("type"):
+                    output.type.CopyFrom(io_types.to_type_proto())
+        if input_types is not None:
+            for input, type in zip(graph.input, input_types):
+                input.type.CopyFrom(type.to_type_proto())
+        if output_types is not None:
+            for output, type in zip(graph.output, output_types):
+                output.type.CopyFrom(type.to_type_proto())
+        if functions is None:
+            sub_functions = self.function_ir.get_called_functions()
+            functions = sub_functions.values()
+        else:
+
+            def to_proto(f):
+                if isinstance(f, onnx.FunctionProto):
+                    return f
+                if isinstance(f, OnnxFunction):
+                    return f.to_function_proto()
+                raise TypeError("Expected a value of type FunctionProto of OnnxFunction")
+
+            functions = [to_proto(f) for f in functions]
+
+        opsets = self.function_ir.opset_imports.copy()
+
+        for proto in functions:
+            if proto.domain not in opsets:
+                opsets[proto.domain] = 1
+            # TODO(rama): Handle conflicts with appropriate error/warning message.
+            for opset in proto.opset_import:
+                if opset.domain not in opsets:
+                    opsets[opset.domain] = opset.version
+
+        if "" not in opsets:
+            # No operator is using the standard opset.
+            # Use the specified version if provided or the default value.
+            opsets[""] = (
+                opset_version if opset_version is not None else onnx.defs.onnx_opset_version()
+            )
+
+        if "ir_version" not in kwargs:
+            kwargs["ir_version"] = select_ir_version(opsets[""])
+        opset_imports = [
+            onnx.helper.make_opsetid(domain, version) for domain, version in opsets.items()
+        ]
+
+        return onnx.helper.make_model(
```

Comment:
Curious why we do this instead of converting from an IR Model?

Reply:
I meant that we can create an ir model first, then serialize it to a model proto. This way we can avoid a copy to the ModelProto when it is initialized from the graph proto.

Reply:
I guess I can append the functions after creating the ModelProto.